src/kernel/cuda/spmm.cuh

```diff
@@ -0,0 +1,272 @@
+/*!
+ *  Copyright (c) 2020 by Contributors
+ * \file kernel/cuda/spmm.cuh
+ * \brief SPMM CUDA kernel function header.
+ */
+#ifndef DGL_KERNEL_CUDA_SPMM_CUH_
+#define DGL_KERNEL_CUDA_SPMM_CUH_
+
+#include "../utils.h"
+#include "../bcast.h"
+#include "atomic.cuh"
+#include "../../runtime/cuda/cuda_common.h"
+
+namespace dgl {
+namespace kernel {
+
+namespace cuda {
+
+template <typename T>
+__device__ __forceinline__ T _ldg(T* addr) {
+#if __CUDA_ARCH__ >= 350
+  return __ldg(addr);
+#else
+  return *addr;
+#endif
+}
+
+template <typename Idx, typename DType,
+          typename BinaryOp, typename ReduceOp,
+          bool UseBcast = false, bool UseIdx = false>
+__global__ void SpMMCooKernel(
+  const DType *ufeat, const DType *efeat, DType *out, Idx *arg_u, Idx *arg_e,
+  const Idx *row, const Idx *col, const Idx* edge_map,
+  int64_t N, int64_t M, int64_t E,
+  int64_t *ubcast_off, int64_t *ebcast_off,
+  int64_t ufeat_len, int64_t efeat_len, int64_t out_len) {
+  // SPMM with COO.
+  Idx ty = blockIdx.y * blockDim.y + threadIdx.y;
+  const Idx stride_y = blockDim.y * gridDim.y;
+  while (ty < E) {
+    const Idx src = _ldg(row + ty);
+    const Idx dst = _ldg(col + ty);
+    const Idx eid = UseIdx ? _ldg(edge_map + ty) : ty;
+    int64_t tx = blockIdx.x * blockDim.x + threadIdx.x;
+    const int64_t stride_x = blockDim.x * gridDim.x;
+    const DType* uoff = BinaryOp::use_lhs ? (ufeat + src * ufeat_len): nullptr;
+    const DType* eoff = BinaryOp::use_rhs ? (efeat + eid * efeat_len): nullptr;
+    DType* outoff = out + dst * out_len;
+    while (tx < out_len) {
+      const int64_t lhs_add = UseBcast ? ubcast_off[tx] : tx;
+      const int64_t rhs_add = UseBcast ? ebcast_off[tx] : tx;
+      DType val = BinaryOp::Call(uoff + lhs_add, eoff + rhs_add);
+      Idx* arguoff = (ReduceOp::require_arg && BinaryOp::use_lhs) ?
+        (arg_u + dst * out_len + tx): nullptr;
+      Idx* argeoff = (ReduceOp::require_arg && BinaryOp::use_rhs) ?
+        (arg_e + dst * out_len + tx): nullptr;
```

Comment:
Is it true that for SpMMCoo, `arguoff` and `argeoff` are always nullptr?

Reply:
I don't understand why this is the case, shouldn't whether `argu/eoff` be depend on the selection of binary operator?

Reply:
But even for max and min reducers, you didn't use the argu and arge buffers right? Instead, you launch another kernel to compute them.

Reply:
You are right.